### PR TITLE
Fix issue in setting the registry for auto logon and resolving the domain user

### PR DESCRIPTION
**Pull request by nigurr**
- [x]  Fix issue in finding the test user session active. Previously we are checking for Active sessions available which doesn't give right result whether test user has active session or not
- [x]  Fix issue in resolving the Domain\Username in Local user groups
- [x]  Fix issue in DTA Process handling if same machine used for v1 and v2

Testing:
- [x] Azure UI Testing - Mseng BID - 3691989
- [x] [DTA v1, v2 2015/2017 UI Testing](https://dtatestingint.visualstudio.com/DTA/_build)

### CODE REVIEW

File: Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1 [hunk 1]
```diff
@@ -547,36 +454,85 @@ namespace MS.VS.TestTools.Config
...
 
-    Set-EnableAutoLogon($domain, $TestUserName, $TestUserPassword)
-    Set-DisableScreenSaverReg
+    Write-Verbose -Message "Test username $TestUser" -Verbose
+    Write-Verbose -Message "Test user domain $Domain" -Verbose
 
-    $isSessionActive = IsAnySessionActive
-    if (-not ($isSessionActive))
+    Set-DisableScreenSaverReg | Out-Null
+    ConfigurePowerOptions -MachineCredential $MachineUserCredential | Out-Null
```
Comment (vimegh) [inline Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1]: >ConfigurePowerOptions -MachineCredential $MachineUserCredential  [](start = 4, length = 64)

ConfigurePowerOptions function doesn't have any param to accept. Please check.

Comment (nigurr) [inline Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1]: Fixed. PS didn't complain

File: Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1 [hunk 2]
```diff
@@ -547,36 +454,85 @@ namespace MS.VS.TestTools.Config
...
+        Remove-ItemProperty -Path 'REGISTRY::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Winlogon' -Name DefaultPassword -ErrorAction SilentlyContinue
+
+        [MS.VS.TestTools.Config.RegHelper]::SetAutoLogonPassword($TestUserPassword)
+    }
+    catch {
+        Write-Warning "Unable to set auto logon: $_"
+    }
+}
+
+# Getting rid of old Session check. WTSEnumerateSessions spits out if there are any active sessions are available
```
Comment (vimegh) [inline Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1]: >are [](start = 100, length = 3)

[opt] not needed

File: Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1 [hunk 3]
```diff
@@ -547,36 +454,85 @@ namespace MS.VS.TestTools.Config
...
+# And we are not checking if the active session belongs to the test user. if not, then the machine should get rebooted
+# Most importantly, WTSEnumerateSessions throws all sort of sessions [Remote PS, RDP, Pings from remote machines]
+# For UI testing, we need active local users logged in instead of sessions check
+function IsTestUserCurrentlyLoggedIn($TestUserName)
+{
+    $computer = $env:COMPUTERNAME
+    try {
+        $quserOut = quser.exe /SERVER:$computer 2>&1
```
Comment (vimegh) [inline Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1]: >quser [](start = 20, length = 5)

Please check windows 7 supports it or not. 
as [https://technet.microsoft.com/en-us/library/cc754583(v=ws.11).aspx](https://technet.microsoft.com/en-us/library/cc754583(v=ws.11).aspx) says
Applies To: Windows Vista, Windows Server 2008, Windows Server 2008 R2, Windows Server 2012, Windows 8

File: Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1 [hunk 4]
```diff
@@ -547,36 +454,85 @@ namespace MS.VS.TestTools.Config
...
+    }
+}
+
+# Getting rid of old Session check. WTSEnumerateSessions spits out if there are any active sessions are available
+# And we are not checking if the active session belongs to the test user. if not, then the machine should get rebooted
+# Most importantly, WTSEnumerateSessions throws all sort of sessions [Remote PS, RDP, Pings from remote machines]
+# For UI testing, we need active local users logged in instead of sessions check
+function IsTestUserCurrentlyLoggedIn($TestUserName)
+{
+    $computer = $env:COMPUTERNAME
```
Comment (vimegh) [inline Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1]: >$computer [](start = 4, length = 9)

This is being used only once. Do we really need it. You can directly pass $env:COMPUTERNAME to quser

File: Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1 [hunk 5]
```diff
@@ -547,36 +454,85 @@ namespace MS.VS.TestTools.Config
...
+        $quserOut = quser.exe /SERVER:$computer 2>&1
+    }
+    catch {
+        return $false
+    }
     
-    Remove-ItemProperty -Path 'REGISTRY::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Winlogon' -Name AutoLogonCount -ErrorAction SilentlyContinue
-    Remove-ItemProperty -Path 'REGISTRY::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Winlogon' -Name DefaultPassword -ErrorAction SilentlyContinue
-    [MS.VS.TestTools.Config.RegHelper]::SetAutoLogonPassword($TestUserPassword)
+    if ($quserOut -match "No user exists")
```
Comment (vimegh) [inline Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1]: >$quserOut -match "No user exists" [](start = 8, length = 33)

will this comparison hold in machine with different locales? please check.

File: Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1 [hunk 6]
```diff
@@ -547,36 +454,85 @@ namespace MS.VS.TestTools.Config
...
+    {
+        Write-Verbose "No users logged in" -Verbose
+        return $false
+    }
+
+    $users = $quserOut -replace '\s{2,}', ',' | ConvertFrom-CSV -Header 'username', 'sessionname', 'id', 'state', 'idleTime', 'logonTime'
```
Comment (vimegh) [inline Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1]: >'username', 'sessionname', 'id', 'state', 'idleTime', 'logonTime' [](start = 72, length = 65)

check for machines with different locales

File: Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1 [hunk 7]
```diff
@@ -547,36 +454,85 @@ namespace MS.VS.TestTools.Config
...
+    $users = $users[1..$users.count]
+
+    foreach ($user in $users)
+    {
+        Write-Verbose "Logged in user: $($user.username) $($user.state)" -Verbose
+        if(($user.username -ieq $TestUserName) -and ($user.state -ieq 'active')) {
```
Comment (vimegh) [inline Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1]: >$user.state -ieq 'active' [](start = 53, length = 25)

please check for different locales

File: Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1 [hunk 8]
```diff
@@ -547,36 +454,85 @@ namespace MS.VS.TestTools.Config
...
+        Set-ItemProperty -Path 'REGISTRY::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Winlogon' -Name DefaultUserName -Value $TestUserName -ErrorAction SilentlyContinue
+        Set-ItemProperty -Path 'REGISTRY::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Winlogon' -Name DefaultDomainName -Value $TestUserDomain -ErrorAction SilentlyContinue
+        Set-ItemProperty -Path 'REGISTRY::HKEY_LOCAL_MACHINE\SOFTWARE\Policies\Microsoft\Windows NT\Reliability' -Name ShutdownReasonOn -Value 0 -ErrorAction SilentlyContinue
+        Set-ItemProperty -Path 'REGISTRY::HKEY_LOCAL_MACHINE\SOFTWARE\Policies\Microsoft\Windows NT\Reliability' -Name ShutdownReasonUI -Value 0 -ErrorAction SilentlyContinue
+        Remove-ItemProperty -Path 'REGISTRY::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Winlogon' -Name AutoLogonCount -ErrorAction SilentlyContinue
+        Remove-ItemProperty -Path 'REGISTRY::HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Winlogon' -Name DefaultPassword -ErrorAction SilentlyContinue
+
+        [MS.VS.TestTools.Config.RegHelper]::SetAutoLogonPassword($TestUserPassword)
+    }
```
Comment (vimegh) [inline Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1]: Add log message -  autologon successfully enabled

File: Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1 (hunk 3 above)
Comment (nigurr) [inline Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1]: We don't support Windows 7 in v2

**Comment by vimegh** [comment]
>     $groupPolicy = (Get-ItemProperty -Path 'REGISTRY::HKEY_CURRENT_USER\Software\Policies\Microsoft\Windows\Control Panel\Desktop' -ErrorAction SilentlyContinue).ScreenSaveActive

Will this point to wow6432node in 64bit machine? 

---
Refers to: Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1:3 in 8ed769f. [](commit_id = 8ed769fc16d589ad23bf43a394f12c164c9aeb56, deletion_comment = False)

File: Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1 (hunk 2 above)
Comment (nigurr) [inline Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1]: didn't get this one. comments/stories are good to have ;)

File: Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1 (hunk 8 above)
Comment (nigurr) [inline Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1]: We print that from C# side for success

File: Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1 [hunk 9]
```diff
@@ -547,36 +454,196 @@ namespace MS.VS.TestTools.Config
...
+                        if (testUserDomain.Equals(domain, StringComparison.OrdinalIgnoreCase) &&
+                            testUsername.Equals(username, StringComparison.OrdinalIgnoreCase) &&
+                            si.State == WTS_CONNECTSTATE_CLASS.WTSActive)
+                        {
+                            WTSFreeMemory(userPtr);
+                            WTSFreeMemory(domainPtr);
+                            return true;
+                        }
+                        WTSFreeMemory(userPtr);
+                        WTSFreeMemory(domainPtr);
```
Comment (vimegh) [inline Tasks/DeployVisualStudioTestAgent/SetupTestMachineForUITests.ps1]: WTS FreeMemory calls should be place in finally

File: Tasks/DeployVisualStudioTestAgent/TestAgentHelper.ps1 [hunk 10]
```diff
@@ -204,5 +204,24 @@ namespace CapabilityHelpers.VisualStudio.Setup.Com
...
-
+function Remove-Service([String] $ServiceName)
+{
+    if(Get-Service $ServiceName -ErrorAction SilentlyContinue)
+    {
+        $service = (Get-WmiObject Win32_Service -filter "name='$ServiceName'")
+        Write-Verbose -Message("Trying to delete service {0}" -f $ServiceName) -Verbose
+        if($service)
+        {
+            $service.StopService()
```
Comment (vimegh) [inline Tasks/DeployVisualStudioTestAgent/TestAgentHelper.ps1]: > $service.StopService() [](start = 11, length = 23)

Please check, what happens, if stopping the service throws an error, as we are not handling that?

Comment (vimegh) [inline Tasks/DeployVisualStudioTestAgent/TestAgentHelper.ps1]: > $service.StopService() [](start = 11, length = 23)

Should we use -Force option here ?

File: Tasks/DeployVisualStudioTestAgent/TestAgentHelper.ps1 [hunk 11]
```diff
@@ -204,5 +204,24 @@ namespace CapabilityHelpers.VisualStudio.Setup.Com
...
+            $deleteServiceCode = $service.Delete()
+            if($deleteServiceCode -eq 0)
```
Comment (vimegh) [inline Tasks/DeployVisualStudioTestAgent/TestAgentHelper.ps1]: >f($deleteServiceCode -eq 0) [](start = 13, length = 27)

As per [https://msdn.microsoft.com/en-us/library/aa389960(VS.85).aspx](https://msdn.microsoft.com/en-us/library/aa389960(VS.85).aspx) Delete method returns 0 if the request was accepted.

**Pull request merged**